homeassistant/helpers/script.py

```diff
@@ -485,26 +513,203 @@ def async_script_timeout(now):
+
+        referenced = set()
+
+        for step in self.sequence:
+            action = _determine_action(step)
+
+            if action == ACTION_CHECK_CONDITION:
+                referenced |= condition.async_extract_devices(step)
+
+            elif action == ACTION_DEVICE_AUTOMATION:
+                referenced.add(step[CONF_DEVICE_ID])
+
+        self._referenced_devices = referenced
+        return referenced
+
+    @property
+    def referenced_entities(self):
+        """Return a set of referenced entities."""
+        if self._referenced_entities is not None:
+            return self._referenced_entities
+
+        referenced = set()
+
+        for step in self.sequence:
+            action = _determine_action(step)
+
+            if action == ACTION_CALL_SERVICE:
+                data = step.get(service.CONF_SERVICE_DATA)
+                if not data:
+                    continue
+
+                entity_ids = data.get(ATTR_ENTITY_ID)
+
+                if entity_ids is None:
+                    continue
+
+                if isinstance(entity_ids, str):
+                    entity_ids = [entity_ids]
+
+                for entity_id in entity_ids:
+                    referenced.add(entity_id)
+
+            elif action == ACTION_CHECK_CONDITION:
+                referenced |= condition.async_extract_entities(step)
+
+            elif action == ACTION_ACTIVATE_SCENE:
+                referenced.add(step[CONF_SCENE])
+
+        self._referenced_entities = referenced
+        return referenced
+
+    def run(self, variables=None, context=None):
+        """Run script."""
+        asyncio.run_coroutine_threadsafe(
+            self.async_run(variables, context), self._hass.loop
+        ).result()
+
+    async def async_run(
+        self, variables: Optional[Sequence] = None, context: Optional[Context] = None
+    ) -> bool:
+        """Run script."""
+        if self.is_running:
+            if self._if_running == IF_RUNNING_IGNORE:
+                self._log("Skipping script")
+                return False
+            if self._if_running == IF_RUNNING_ERROR:
+                self._raise("Already running")
+            if self._if_running == IF_RUNNING_RESTART:
+                self._log("Restarting script")
+                self.async_stop()
+
+        self.last_triggered = utcnow()
+        if self._run_mode == RUN_MODE_LEGACY:
+            if self.is_running:
+                shared = cast(Optional[_LegacyScriptRun], self._runs[0])
+            else:
+                shared = None
+            run = cast(
+                _ScriptRunBase,
+                _LegacyScriptRun(self._hass, self, variables, context, shared),
+            )
+            self._runs.append(run)
+            await run.async_run()
+        else:
+            run = cast(_ScriptRunBase, _ScriptRun(self._hass, self, variables, context))
+            self._runs.append(run)
+            cast(_ScriptRun, run).task = self._hass.async_create_task(run.async_run())
```

Comment:
This seems very weird. I would expect `run.async_run()` would just set its task with `self.task = asyncio.current_task()`

Reply:
Ah! I looked and Googled and ... trying to find `asyncio.current_task()`. THANK YOU!! (RTOS's I've used had a call like that, but I couldn't find an equivalent in `asyncio`.) That should certainly simplify things. Let me work on that, and if I can simplify, I'll add a new commit.